server/src/services/oanda.py

```diff
@@ -0,0 +1,57 @@
+import requests_async as requests
+from datetime import datetime, timedelta
+from decimal import *
+
+from src.settings import settings
+from src.utils.logging import structlog
+
+logger = structlog.getLogger(__name__)
+
+
+def create_base_headers():
+    return {
+        "Content-Type": "application/json; charset=UTF-8",
+        "Accept": "application/json; charset=UTF-8",
+        "Authorization": f"Bearer {settings.oanda_api_key}"
+    }
+
+
+async def create_order(pair, direction, price):
+    r = await requests.post(
+        f"{settings.oanda_base_url}/v3/accounts/{settings.oanda_account_id}/orders",
+        json=create_open_order_request_body(pair, direction, price),
+        headers=create_base_headers(),
+    )
+    body = r.json()
+
+    try:
+        return body["orderCreateTransaction"]["id"]
+    except:
+        logger.error(
+            "Error placing order",
+            response_body=body
+        )
+
+
+def create_open_order_request_body(pair, direction, price):
+    instrument = f"{pair[:3]}_{pair[3:]}"
+    price_str = str(price)
+    good_till_date = (datetime.now() + timedelta(seconds=10)).timestamp()
+    negative_decimal_places = Decimal(price_str).as_tuple().exponent
+    take_profit_distance = 2 * 10 ** negative_decimal_places
```

Comment:
it turns out we want our distance to be 2 x the smallest unit possible for the instrument. To do this we grab the amount of decimal places for our price, returned as a negative number, and calculate from there.

My question here is, do we need to know the amount of decimal places for a given asset ahead of time (i.e in a map) or will we _always_ receive a price to the maximum accuracy of an instrument, even if we make a EUR trade for say 1.06000. To the logs me thinks.

Reply:
UPDATE: A quick glance through the logs suggets that we _do_ receive prices without trailing 0s, so we will have to know the maximum decimal places of a given instrument ahead of time.